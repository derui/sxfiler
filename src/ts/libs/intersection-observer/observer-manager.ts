// define application-global intersection observer management.
import uuid from "uuid";
import "intersection-observer";

export type RootId = string;
export type IntersectionChandedHandler = (entry: IntersectionObserverEntry) => void;

export interface IntersectionTarget {
  element: Element;
  handler: IntersectionChandedHandler;
}

interface Observer {
  instance: IntersectionObserver | undefined;
  targets: Map<Element, IntersectionTarget>;
}

const intersectionObserverMap = new Map<RootId, Observer>();

// get observer binded rootId
function getObserver(rootId: RootId): Observer | undefined {
  return intersectionObserverMap.get(rootId);
}

export function observingTargets(rootId: RootId): IntersectionTarget[] | undefined {
  const observer = intersectionObserverMap.get(rootId);

  if (!observer) {
    return undefined;
  }

  return Array.from(observer.targets.values());
}

// observe the element by root
export function observe(rootId: RootId, target: IntersectionTarget) {
  const observer = getObserver(rootId);

  if (!observer) {
    throw new Error(`Not found root instance by ${rootId}, try "createRoot" first.`);
  }

  observer.targets.set(target.element, target);

  const { instance } = observer;
  if (instance) {
    instance.observe(target.element);
  }
}

// unobserve the element from root
export function unobserve(target: IntersectionTarget): boolean {
  let deleted = false;

  for (let entry of intersectionObserverMap.entries()) {
    let [_, observer] = entry;
    if (observer.instance) {
      observer.instance.unobserve(target.element);
    }
    deleted = observer.targets.delete(target.element) || deleted;
  }

  return deleted;
}

// create and register new root
export function createRoot(): RootId {
  const id = uuid.v4();

  intersectionObserverMap.set(id, { instance: undefined, targets: new Map() });

  return id;
}

export function deleteRoot(rootId: RootId): boolean {
  const observer = getObserver(rootId);

  if (!observer) {
    return false;
  }

  intersectionObserverMap.delete(rootId);

  const { instance, targets } = observer;
  if (instance) {
    targets.forEach(e => instance.unobserve(e.element));
    instance.disconnect();
  }

  return true;
}

function callback(rootId: RootId): (entries: IntersectionObserverEntry[]) => void {
  return entries => {
    const observer = getObserver(rootId);

    if (!observer) {
      throw new Error(`Not found root instance of ${rootId}`);
    }

    entries.forEach(entry => {
      const target = observer.targets.get(entry.target);
      if (target) {
        target.handler(entry);
      }
    });
  };
}

export function registerObserverRoot(rootId: RootId, option: IntersectionObserverInit): IntersectionObserver {
  const observer = getObserver(rootId);

  if (!observer) {
    throw new Error(`Not found root ${rootId}`);
  }

  const { instance, targets } = observer;
  if (observer && instance && instance.root !== option.root) {
    targets.forEach(e => instance.unobserve(e.element));
    instance.disconnect();
    observer.instance = undefined;
  }

  const newInstance = new IntersectionObserver(callback(rootId), option);
  observer.instance = newInstance;
  targets.forEach(e => newInstance.observe(e.element));
  return newInstance;
}
