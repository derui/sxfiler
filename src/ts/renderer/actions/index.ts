import * as directoryM from './directory';
import * as paneOperationM from './pane-operations';
import * as windowM from './window';

export const directory = directoryM;

export const paneOperation = paneOperationM;

export const window = windowM;
