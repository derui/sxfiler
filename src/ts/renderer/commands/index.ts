import * as Directory from './defaults/directory';
import * as Pane from './defaults/pane';
import * as Window from './defaults/window';

// export default commands.
import * as types from './types';

// exporting all types from current module.
export * from './types';

export const defaultCommands: types.Commands = {
  commands: [
    Directory.upDirectory,
    Directory.downDirectory,
    Pane.upCursor,
    Pane.downCursor,
    Pane.changePane,
    Window.quitApplication
  ]
};
