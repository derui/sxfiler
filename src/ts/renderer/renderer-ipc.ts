import R from 'ramda';
import Bus from 'sxfiler/renderer/utils/bus';
import {IPC_KEYS} from 'sxfiler/common/constants';
import uuid from 'uuid';

interface Event {
  sender: {
    send: (event: string, ...args: any[]) => void;
  };
}

export interface RendererIPC {
  /**
   * Send a message to channel with args.
   *
   * Notice: When you send message from this, you must take channel that have 'REQUEST_' prefix.
   *
   * Return an observeable sequence that only subscribe event once. You should subscribe it or
   * map/connect/concat... any observeable operations with any other sequence what you want.
   *
   * @param {string} channel channel for sending message
   * @param {...any} args message to send channel
   * @return {Rx.Observerable} An observeable sequence only subscribe once.
   */
  send(channel: string, ...args: any[]): Rx.Observable<any>;

  /**
   * Subscribe observer to channel bus.
   *
   * Using this method if you subscribe all values streaming in specified channel.
   * You do not forget to call dispose from returning disposable.
   *
   * @param {string} channel channel for subscribing observer
   * @param {Rx.Observer} observer observer to subscribe for channel
   * @return {Rx.Disposable} The disposable
   */
  subscribe(channel: string, observer: (val: any) => void): Rx.Disposable;
}

/**
 * IPC event management for Renderer process.
 *
 * This class is based on {Rx.Observable} to publish event handling, so it only
 * provide `subscribe` and `send` method.
 */
export default class RendererIPCImpl implements RendererIPC {

  private _ipc: any;
  private _bus: Bus = new Bus();

  /**
   * Construct RendererIPCHandler
   *
   * Notice for parameter ipc has to require from provided 'require' from 'Electron'.
   *
   * @param {!IPC} ipc ipc module using this
   */
  constructor(ipc: any) {
    this._ipc = ipc;

    this._ipc.on(IPC_KEYS.FINISH_FILES_IN_DIRECTORY, this._onFinishFilesInDirectory.bind(this));
  }

  /**
   * Send a message to channel with args.
   *
   * Notice: When you send message from this, you must take channel that have 'REQUEST_' prefix.
   *
   * Return an observeable sequence that only subscribe event once. You should subscribe it or
   * map/connect/concat... any observeable operations with any other sequence what you want.
   *
   * @param {string} channel channel for sending message
   * @param {...any} args message to send channel
   * @return {Rx.Observerable} An observeable sequence only subscribe once.
   */
  public send(channel: string, ...args: any[]): Rx.Observable<any> {
    let currentId: string = uuid.v4();
    this._ipc.send(channel, ...args.concat([currentId]));

    return this._bus.bus(channel).filter((_args: any[]) => R.last(_args) === currentId).first();
  }

  /**
   * Subscribe observer to channel bus.
   *
   * Using this method if you subscribe all values streaming in specified channel.
   * You do not forget to call dispose from returning disposable.
   *
   * @param {string} channel channel for subscribing observer
   * @param {Rx.Observer} observer observer to subscribe for channel
   * @return {Rx.Disposable} The disposable
   */
  public subscribe(channel: string, observer: (val: any) => void): Rx.Disposable {
    return this._bus.bus(channel).subscribe(observer);
  }

  /**
   * @param {event} ev Event object from ipc event emitter
   * @param {object} err error information for requesting file informations
   * @param {string} path An absolute path what traversed
   * @param {array[object]} files fs.Stats objects in requested directory
   * @param {Pane} pane side of sending request
   * @param {string} uuid The unique event id
   */
  private _onFinishFilesInDirectory(ev: Event, ...args: any[]): void {
    let bus: Rx.Subject<any> = this._bus.bus(IPC_KEYS.REQUEST_FILES_IN_DIRECTORY);

    bus.onNext(args);
  }
}
