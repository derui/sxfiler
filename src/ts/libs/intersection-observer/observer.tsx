import * as React from "react";
import * as observer from "./observer-manager";

export interface Props {
  rootId: string;
  onChange?: (entry: IntersectionObserverEntry, unobserve: () => void) => void;
}

export class Component extends React.Component<Props> {
  private currentTarget: observer.IntersectionTarget | null = null;

  componentDidMount() {
    this.observe();
  }

  // unobserve when unmount
  componentWillUnmount() {
    this.unobserve();
  }

  componentDidUpdate(prevProps: Props) {
    if (this.currentTarget) {
      let changed = false;
      if (prevProps.rootId !== this.props.rootId || prevProps.onChange !== this.props.onChange) {
        this.unobserve();
        changed = true;
      }

      if (changed) {
        this.observe();
      }
    }
  }

  observe = () => {
    if (!this.currentTarget) {
      return;
    }

    observer.observe(this.props.rootId, this.currentTarget);
  };

  unobserve = () => {
    if (this.currentTarget) {
      observer.unobserve(this.currentTarget);
    }
  };

  // call handler if changed the state of intersection
  handleChange = (entry: IntersectionObserverEntry) => {
    if (this.props.onChange) {
      this.props.onChange(entry, this.unobserve);
    }
  };

  // observing new element
  handleNode = (element: Element | null) => {
    if (this.currentTarget) {
      this.unobserve();
    }

    if (!element) {
      return;
    }

    this.currentTarget = {
      element,
      handler: this.handleChange,
    };
  };

  render() {
    const child = React.Children.only(this.props.children);
    if (!React.isValidElement(child)) {
      return null;
    }

    return React.cloneElement(child as React.ReactElement<any>, {
      ref: this.handleNode,
    });
  }
}
