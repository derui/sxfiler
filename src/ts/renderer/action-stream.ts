import * as Rx from 'rx';
import {EventEmitter} from 'events';
import {RendererIPC} from './renderer-ipc';

type actionArgument = (ipc: RendererIPC) => Rx.Observable<any> | Rx.Observable<any>;

/**
 * An interface of ActionStream
 */
export interface ActionStream {
  /**
   * Publish action to action stream.
   *
   * @param {function(ipc:RendererIPC):Rx.Observable|Rx.Observable} act - an action to publish
   */
  doAction(act: actionArgument): void;

  /**
   * Subscribe events from this.
   *
   * @param {function(value: any)|Rx.Observer<any>} onNext - an onNext callback or observer
   * @param {function(exception: any)} onError - an onError callback
   * @param {function()} onCompleted - an onCompleted callback
   * @return {Rx.IDisposable} a disposable
   */
  subscribe(onNext?: ((value: any) => void|Rx.Observer<any>), onError?: (exception: any) => void, onCompleted?: () => void)
  : Rx.IDisposable;

  /**
   * Dispose published stream from this.
   */
  dispose(): void;
}

/**
 * ActionStream provide action emitter and method to subscribe it.
 */
export default class ActionStreamImpl implements ActionStream {
  private _e: NodeJS.EventEmitter;
  private _broadcastingObservable: Rx.Observable<any>;
  private _broadcastingDisposable: Rx.IDisposable;

  public constructor(ipc: RendererIPC) {
    this._e = new EventEmitter();

    // define observable from EventEmitter
    this._broadcastingObservable = Rx.Observable.fromEventPattern(
      (h: (data: any) => any) => this._e.addListener('action', h),
      (h: (data: any) => any) => this._e.removeListener('action', h),
      (act: any) => {
        if (typeof act === 'function') {
          return (<actionArgument>act)(ipc);
        } else {
          return act;
        }
      }
    ).flatMap((x: Rx.Observable<any>) => x).publish();

    this._broadcastingDisposable = this._broadcastingObservable.connect();
  }

  /**
   * Publish action to action stream.
   *
   * @param {function(ipc:RendererIPC):Rx.Observable|Rx.Observable} act - an action to publish
   */
  public doAction(act: actionArgument): void {
    this._e.emit('action', act);
  }

  /**
   * Subscribe events from this.
   *
   * @param {function(value: any)|Rx.Observer<any>} onNext - an onNext callback or observer
   * @param {function(exception: any)} onError - an onError callback
   * @param {function()} onCompleted - an onCompleted callback
   * @return {Rx.IDisposable} a disposable
   */
  public subscribe(onNext?: ((value: any) => void|Rx.Observer<any>), onError?: (exception: any) => void, onCompleted?: () => void)
  : Rx.IDisposable {
    return this._broadcastingObservable.subscribe(onNext, onError, onCompleted);
  }

  /**
   * Dispose published stream from this.
   */
  public dispose(): void {
    if (this._broadcastingDisposable) {
      this._broadcastingDisposable.dispose();
    }
    this._broadcastingDisposable = null;
  }
}
