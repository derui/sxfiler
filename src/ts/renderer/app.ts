/// <reference path="../../../typings/tsd.d.ts" />
import * as Rx from 'rx';

import * as ReactDOM from 'react-dom';
import * as React from 'react';
import uuid from 'uuid';
import R from 'ramda';

import Main from './components/main';
import ActionStream from './action-stream';
import RendererIPC from './renderer-ipc';
import {PANE} from 'sxfiler/common/constants';
import Directory from './binders/directory';
import CurrentPane from './binders/current-pane';
import * as types from 'sxfiler/renderer/types';
import * as actions from './actions';
import KeyHandler from './key-handlers';
import {convertEventToDescriptor} from 'sxfiler/renderer/utils/key-binding-util';

let ipc = new RendererIPC(window.require('electron').ipcRenderer);
let actionStream = new ActionStream(ipc);

let currentPane = new CurrentPane(PANE.LEFT, actionStream);
let leftPane = new Directory(PANE.LEFT, actionStream);
let rightPane = new Directory(PANE.RIGHT, actionStream);
let paneInfo = Rx.Observable.combineLatest(
  leftPane.store.getObservable(),
  rightPane.store.getObservable(),
  (left: types.PaneInfo, right: types.PaneInfo): types.PaneInfos => {
    return {
      [PANE.LEFT]: left,
      [PANE.RIGHT]: right
    };
  });

let globalState = Rx.Observable.combineLatest(
  paneInfo,
  currentPane.store.getObservable(),
  (paneInfo: types.PaneInfos, currentPane: string): types.GlobalState => {
    return {
      paneInfo,
      currentPane
    };
  }).publish();

globalState.connect();

globalState.subscribe((state: types.GlobalState) => {
  ReactDOM.render(React.createElement(Main, {state, actionStream}), document.querySelector('.entry'));
});

// setup keyevent stream
interface StateWithEvent {
  event: KeyboardEvent;
  eventId: string;
  state: types.GlobalState;
};

let keyEvents = Rx.Observable.combineLatest(
  Rx.Observable.fromEvent(document.body, 'keydown')
    .select((e: KeyboardEvent) => {
      return {
        event: e,
        id: uuid.v4()
      };
    }),
  globalState,
  (ev: {event: KeyboardEvent, id: string}, state: types.GlobalState): StateWithEvent => {
    return {
      event: ev.event,
      eventId: ev.id,
      state
    };
  });

// this variable only store event id from observable sequence as `keyEvents'
let _lastEventId = '';
let keyHandler = new KeyHandler();
keyEvents.subscribe((state: StateWithEvent) => {
  if (state.eventId === _lastEventId) {
    return;
  }

  _lastEventId = state.eventId;

  let _e = R.clone(state.event);
  let desc = convertEventToDescriptor(_e);
  actionStream.doAction(
    keyHandler.invokeCommand(desc, state.state)
  );
});

actionStream.doAction(actions.directory.getDirectory('.', PANE.LEFT));
actionStream.doAction(actions.directory.getDirectory('.', PANE.RIGHT));
