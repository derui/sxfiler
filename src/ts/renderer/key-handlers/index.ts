// provide key handler with key bindings
import * as Rx from 'rx';
import R from 'ramda';
import * as types from 'sxfiler/common/types';
import {GlobalState} from 'sxfiler/renderer/types';
import * as commands from 'sxfiler/renderer/commands';
import {getKeyIdentifier, bindMappingWithCommands} from 'sxfiler/renderer/utils/key-binding-util';
import defaultMapping from './defaults';
import * as actions from 'sxfiler/renderer/actions';
import * as Maybe from 'sxfiler/common/maybe';

let defaultCommands = commands.defaultCommands;

/**
 * Merge two keyDescriptors to one keyDescriptors.
 *
 * @param {array[KeyDescriptor]} src1 key bindings to merge with src2
 * @param {array[KeyDescriptor]} src2 key bindings to merge with src1
 * @return {array[KeyDescriptor]} merged key descriptors
 */
function mergeDescriptors(src1: types.KeyDescriptor[], src2: types.KeyDescriptor[]): types.KeyDescriptor[] {
  'use strict';

  let toObj: any = R.reduce((memo: any, v: types.KeyDescriptor) => R.merge(memo, {[getKeyIdentifier(v)]: v}), {});
  let src1Descs = toObj(src1), src2Descs = toObj(src2);

  return R.values(R.merge(src1Descs, src2Descs));
}

export interface KeyHandler {
  /**
   * Invoke a command binded a key-combination which calculated from KeyboardEvent.
   * @param {KeyDescriptor} desc - a descriptor to invoke command
   * @param {IGlobalState} globalState - the global state to pass command
   * @return {Rx.Observable} result of command
   */
  invokeCommand(desc: types.KeyDescriptor, globalState: GlobalState): Rx.Observable<any>;
}

/**
 * KeyHandler provide match keyboard input with keybindings that are defined user or default,
 *  and run command in it.
 */
export default class KeyHandlerImpl implements KeyHandler {
  private keyBindings: types.KeyBindings<commands.Command>;

  /**
   * Construct KeyHandler.
   *
   * @param {IKeyBindings} [keyBindings] key bindings to append to default binding
   */
  constructor(keyBindings?: types.KeyBindings<commands.Command>) {
    this.keyBindings = bindMappingWithCommands(defaultMapping, defaultCommands);

    // if keybindings given, merge with default bindings.
    if (keyBindings) {
      this.keyBindings.bindings = mergeDescriptors(this.keyBindings.bindings, keyBindings.bindings);
      this.keyBindings.mapping = R.merge(this.keyBindings.bindings, keyBindings.mapping);
    }
  }

  /**
   * Invoke a command binded a key-combination which calculated from KeyboardEvent.
   * @param {KeyDescriptor} desc - a descriptor to invoke command
   * @param {IGlobalState} globalState - the global state to pass command
   * @return {Rx.Observable} result of command
   */
  public invokeCommand(desc: types.KeyDescriptor, globalState: GlobalState): Rx.Observable<any> {

    let command = this.matchKey(desc);
    if (Maybe.isNone(command)) {
      return Rx.Observable.empty();
    }

    return command.val.command(globalState, actions);
  }

  /**
   * Match given KeyboardEvent with keybindings in this class
   *
   * @param {KeyDescriptor} desc - a descriptor to invoke command
   * @return {ICommand} a command if match keyboard event with keybindings. return null if not.
   */
  private matchKey(desc: types.KeyDescriptor): Maybe.Data<commands.Command> {
    if (!desc) {
      return Maybe.NONE;
    }

    let id = getKeyIdentifier(desc);

    let binding = this.keyBindings.mapping[id];
    if (!binding) {
      return Maybe.NONE;
    }
    return Maybe.just(binding);
  }
}
