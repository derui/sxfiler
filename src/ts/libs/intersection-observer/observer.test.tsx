import * as React from "react";
import * as enzyme from "enzyme";
import renderer from "react-test-renderer";

import * as observer from "./observer-manager";
import * as Observer from "./observer";

afterEach(() => {
  jest.restoreAllMocks();
});

describe("Intersection Observer for React", () => {
  describe("Observer Element", () => {
    it("throws if component has more than a child", () => {
      const component = (
        <Observer.Component rootId="tmp">
          <span />
          <span />
        </Observer.Component>
      );

      expect(() => renderer.create(component)).toThrowErrorMatchingSnapshot();
    });

    it("observe when component mounted", () => {
      const id = observer.createRoot();
      const instance = observer.registerObserverRoot(id, {
        root: document.createElement("div"),
      });
      const spy = jest.spyOn(instance, "observe");

      const wrapper = enzyme.mount(
        <Observer.Component rootId={id}>
          <span>foo</span>
        </Observer.Component>
      );

      expect(spy.mock.calls).toHaveLength(1);
      expect(wrapper.find("span")).toHaveLength(1);
    });

    it("unobserve when component unmounted", () => {
      const id = observer.createRoot();
      observer.registerObserverRoot(id, {
        root: document.createElement("div"),
      });

      const wrapper = enzyme.mount(
        <Observer.Component rootId={id}>
          <span>foo</span>
        </Observer.Component>
      );
      const spy = jest.spyOn(wrapper.instance() as any, "unobserve");
      wrapper.unmount();

      expect(spy).toHaveBeenCalledTimes(2);
    });

    describe("update", () => {
      it("reobserve if root changed", () => {
        const id = observer.createRoot();
        const id2 = observer.createRoot();
        observer.registerObserverRoot(id, {
          root: document.createElement("div"),
        });
        const handler = jest.fn();

        const wrapper = enzyme.mount(
          <Observer.Component rootId={id} onChange={handler}>
            <span>foo</span>
          </Observer.Component>
        );

        const observe = jest.spyOn(wrapper.instance() as any, "observe");
        const unobserve = jest.spyOn(wrapper.instance() as any, "unobserve");

        wrapper.setProps({ rootId: id2 });

        expect(observe).toHaveBeenCalledTimes(1);
        expect(unobserve).toHaveBeenCalledTimes(1);
      });

      it("reobserve if handler changed", () => {
        const id = observer.createRoot();
        observer.registerObserverRoot(id, {
          root: document.createElement("div"),
        });

        const wrapper = enzyme.mount(
          <Observer.Component rootId={id}>
            <span>foo</span>
          </Observer.Component>
        );

        const observe = jest.spyOn(wrapper.instance() as any, "observe");
        const unobserve = jest.spyOn(wrapper.instance() as any, "unobserve");

        wrapper.setProps({ rootId: id, onChange: () => {} });

        expect(observe).toHaveBeenCalledTimes(1);
        expect(unobserve).toHaveBeenCalledTimes(1);
      });
    });
  });
});
