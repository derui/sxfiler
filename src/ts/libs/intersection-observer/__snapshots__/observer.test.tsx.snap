// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Intersection Observer for React Observer Element throws if component has more than a child 1`] = `"React.Children.only expected to receive a single React element child."`;
